### ISSUE TITLE
Wrong host is used for ansible_host variable expansion with delegate_to

### ISSUE THREAD

**Issue by DaDummy** [ansible/ansible#76868]
### Summary

_Edit: The issue appears to affect other modules aswell._

When using `ansible-builtin.slurp` with `delegate_to` where the original host has `ansible_host` configured in `host_vars` to `"{{ lan_ipv4 }}"` and the delegated host has not, `slurp` tries to connect using `ansible_host` from the delegated host, but resolves it using variables from the original host, leading to erratic behavior.

In other words I expect the following from `slurp` as well as any other module when executing the first task of the play below for the host `green.dashtec.net`:
1. `delegate_to: "ci-dind"` is evaluated: the string `"ci-dind"` is expanded using the `hostvars` of `green.dashtec.net`, but since it does not contain any variables, it just evaluates to `"ci-dind"`
2. `"ci-dind"` is looked up in the inventory, found and used
3. When connecting, `ansible_host: "{{ lan_ipv4 }}"` of `ci-dind` is expanded using the `hostvars` of `ci-dind`: `lan_ipv4: 192.168.111.10`
4. Ansible attempts to connect to 192.168.111.10

What happens instead is this:
1. `delegate_to: "ci-dind"` is evaluated: the string `"ci-dind"` is expanded using the `hostvars` of `green.dashtec.net`, but since it does not contain any variables, it just evaluates to `"ci-dind"` (as expected)
2. `"ci-dind"` is looked up in the inventory, found and used (as expected)
3. When connecting, `ansible_host: "{{ lan_ipv4 }}"` of `ci-dind` is expanded using the `hostvars` of `green.dashtec.net`: `lan_ipv4: 192.168.112.11` (`ansible_host` is expanded using the wrong variables)
4. Ansible attempts to connect to 192.168.112.11 (connection is attempted to a host that is not in the inventory)

### Issue Type

Bug Report

### Component Name

lib/ansible/modules/slurp.py

### Ansible Version

```console
$ ansible --version
ansible [core 2.12.1]
  config file = /home/chris/work/ps/woe/infrastructure/ansible.cfg
  configured module search path = ['/home/chris/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3/dist-packages/ansible
  ansible collection location = /home/chris/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible
  python version = 3.9.7 (default, Sep 10 2021, 14:59:43) [GCC 11.2.0]
  jinja version = 2.11.3
  libyaml = True
```

### Configuration

```console
$ ansible-config dump --only-changed
DEFAULT_HOST_LIST(<playbook_path>/ansible.cfg) = ['<playbook_path>/inventory/hosts.ini']
DEFAULT_REMOTE_USER(env: ANSIBLE_REMOTE_USER) = username
DEFAULT_STDOUT_CALLBACK(<playbook_path>/ansible.cfg) = yaml
ERROR_ON_MISSING_HANDLER(<playbook_path>/ansible.cfg) = True
INTERPRETER_PYTHON(<playbook_path>/ansible.cfg) = python3
RETRY_FILES_ENABLED(<playbook_path>/ansible.cfg) = False
```

### OS / Environment

Ansible controller:
```
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 21.10
Release:        21.10
Codename:       impish
```

Hosts:
```
$ lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 10 (buster)
Release:        10
Codename:       buster
```

### Steps to Reproduce

```yaml (paste below)
- hosts: green.dashtec.net,ci-dind

  gather_facts: no

  tasks:
    - name: Retrieve file
      delegate_to: "ci-dind"
      become: yes
      ansible.builtin.slurp:
        src: /etc/hostname
      register: slurped

    - debug:
        msg: "{{ ansible_host }} - {{ lan_ipv4 }} - {{ slurped.content | b64decode | trim }}"
```

### Expected Results

```
TASK [debug] **********************************************************************************************************
ok: [green.dashtec.net] => 
  msg: green.dashtec.net - 192.168.112.11 - ci-dind
ok: [ci-dind] => 
  msg: 192.168.111.10 - 192.168.111.10 - ci-dind
```
The expectation is that when executing that task for `green.dashtec.net`, `slurp` connects to `ci-dind` and retrieves the file's contents.

### Actual Results

```console
ansible-playbook [core 2.12.1]
  config file = <playbook_path>/ansible.cfg
  configured module search path = ['/home/chris/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3/dist-packages/ansible
  ansible collection location = /home/chris/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible-playbook
  python version = 3.9.7 (default, Sep 10 2021, 14:59:43) [GCC 11.2.0]
  jinja version = 2.11.3
  libyaml = True
Using <playbook_path>/ansible.cfg as config file
host_list declined parsing <playbook_path>/inventory/hosts.ini as it did not pass its verify_file() method
script declined parsing <playbook_path>/inventory/hosts.ini as it did not pass its verify_file() method
auto declined parsing <playbook_path>/inventory/hosts.ini as it did not pass its verify_file() method
yaml declined parsing <playbook_path>/inventory/hosts.ini as it did not pass its verify_file() method
Parsed <playbook_path>/inventory/hosts.ini inventory source with ini plugin
redirecting (type: callback) ansible.builtin.yaml to community.general.yaml
redirecting (type: callback) ansible.builtin.yaml to community.general.yaml
Skipping callback 'default', as we already have a stdout callback.
Skipping callback 'minimal', as we already have a stdout callback.
Skipping callback 'oneline', as we already have a stdout callback.

PLAYBOOK: bug.yml **************************************************************
1 plays in bug.yml

PLAY [green.dashtec.net,ci-dind] ***********************************************
META: ran handlers

TASK [Retrieve file] ***********************************************************
task path: <playbook_path>/bug.yml:6
<192.168.112.11> ESTABLISH SSH CONNECTION FOR USER: username
<192.168.112.11> SSH: EXEC ssh -F ./ssh.cfg -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="username"' -o ConnectTimeout=10 192.168.112.11 '/bin/sh -c '"'"'echo ~username && sleep 0'"'"''
<192.168.111.10> ESTABLISH SSH CONNECTION FOR USER: username
<192.168.111.10> SSH: EXEC ssh -F ./ssh.cfg -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="username"' -o ConnectTimeout=10 192.168.111.10 '/bin/sh -c '"'"'echo ~username && sleep 0'"'"''
<192.168.111.10> (0, b'/home/username\n', b'')
<192.168.111.10> ESTABLISH SSH CONNECTION FOR USER: username
<192.168.111.10> SSH: EXEC ssh -F ./ssh.cfg -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="username"' -o ConnectTimeout=10 192.168.111.10 '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo /home/username/.ansible/tmp `"&& mkdir "` echo /home/username/.ansible/tmp/ansible-tmp-1643320411.7252398-78432-265912787891396 `" && echo ansible-tmp-1643320411.7252398-78432-265912787891396="` echo /home/username/.ansible/tmp/ansible-tmp-1643320411.7252398-78432-265912787891396 `" ) && sleep 0'"'"''
<192.168.111.10> (0, b'ansible-tmp-1643320411.7252398-78432-265912787891396=/home/username/.ansible/tmp/ansible-tmp-1643320411.7252398-78432-265912787891396\n', b'')
Using module file /usr/lib/python3/dist-packages/ansible/modules/slurp.py
<192.168.111.10> PUT /home/chris/.ansible/tmp/ansible-local-78427pmvs_vo2/tmp9m0u7yjg TO /home/username/.ansible/tmp/ansible-tmp-1643320411.7252398-78432-265912787891396/AnsiballZ_slurp.py
<192.168.111.10> SSH: EXEC sftp -b - -F ./ssh.cfg -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="username"' -o ConnectTimeout=10 '[192.168.111.10]'
<192.168.111.10> (0, b'sftp> put /home/chris/.ansible/tmp/ansible-local-78427pmvs_vo2/tmp9m0u7yjg /home/username/.ansible/tmp/ansible-tmp-1643320411.7252398-78432-265912787891396/AnsiballZ_slurp.py\n', b'')
<192.168.111.10> ESTABLISH SSH CONNECTION FOR USER: username
<192.168.111.10> SSH: EXEC ssh -F ./ssh.cfg -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="username"' -o ConnectTimeout=10 192.168.111.10 '/bin/sh -c '"'"'chmod u+x /home/username/.ansible/tmp/ansible-tmp-1643320411.7252398-78432-265912787891396/ /home/username/.ansible/tmp/ansible-tmp-1643320411.7252398-78432-265912787891396/AnsiballZ_slurp.py && sleep 0'"'"''
<192.168.111.10> (0, b'', b'')
<192.168.111.10> ESTABLISH SSH CONNECTION FOR USER: username
<192.168.111.10> SSH: EXEC ssh -F ./ssh.cfg -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="username"' -o ConnectTimeout=10 -tt 192.168.111.10 '/bin/sh -c '"'"'sudo -H -S -n  -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-phqqxvbigswiohwvpwzznysrtsjrfles ; python3 /home/username/.ansible/tmp/ansible-tmp-1643320411.7252398-78432-265912787891396/AnsiballZ_slurp.py'"'"'"'"'"'"'"'"' && sleep 0'"'"''
Escalation succeeded
<192.168.111.10> (0, b'\r\n{"content": "Y2ktZGluZAo=", "source": "/etc/hostname", "encoding": "base64", "invocation": {"module_args": {"src": "/etc/hostname"}}}\r\n', b'Shared connection to 192.168.111.10 closed.\r\n')
<192.168.111.10> ESTABLISH SSH CONNECTION FOR USER: username
<192.168.111.10> SSH: EXEC ssh -F ./ssh.cfg -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o 'User="username"' -o ConnectTimeout=10 192.168.111.10 '/bin/sh -c '"'"'rm -f -r /home/username/.ansible/tmp/ansible-tmp-1643320411.7252398-78432-265912787891396/ > /dev/null 2>&1 && sleep 0'"'"''
<192.168.111.10> (0, b'', b'')
ok: [ci-dind] => changed=false 
  content: Y2ktZGluZAo=
  encoding: base64
  invocation:
    module_args:
      src: /etc/hostname
  source: /etc/hostname
<192.168.112.11> (255, b'', b'kex_exchange_identification: Connection closed by remote host\r\nConnection closed by UNKNOWN port 65535\r\n')
fatal: [green.dashtec.net -> ci-dind]: UNREACHABLE! => changed=false 
  msg: |-
    Failed to connect to the host via ssh: kex_exchange_identification: Connection closed by remote host
    Connection closed by UNKNOWN port 65535
  unreachable: true

TASK [debug] *******************************************************************
task path: <playbook_path>/bug.yml:13
ok: [ci-dind] => 
  msg: 192.168.111.10 - 192.168.111.10 - ci-dind
META: ran handlers
META: ran handlers

PLAY RECAP *********************************************************************
ci-dind                    : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
green.dashtec.net          : ok=0    changed=0    unreachable=1    failed=0    skipped=0    rescued=0    ignored=0
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

**Comment by DaDummy** [comment]
Slurp did behave correctly in a previous version, though I am unsure which version exactly that was.

**Comment by DaDummy** [comment]
I just tried to workaround the issue by using the `command` module like this:

```yaml
    - name: Retrieve file
      delegate_to: "ci-dind"
      become: yes
      command: cat /etc/hostname
      register: slurped
```

But it yields the same erratic behavior. This regression might be more severe than I expected. :/

**Comment by DaDummy** [comment]
Found a workaround:
At the start of any affected play, add this task:
```yaml
    - name: Workaround ansible #76868
      set_fact:
        ansible_host: "{{ ansible_host }}"
```

This forces ansible to resolve `ansible_host` to a static string right then and there.

This will obviously break any black magic involving changing variables/facts `ansible_host` depends on during a play. But noone does that, right? Right???

**Comment by s-hertel** [comment]
@DaDummy Can you share the inventory and host vars files to complete the reproducer?

**Comment by DaDummy** [comment]
@s-hertel a `hosts.ini` like this should be enough for reproduction:
```
green.dashtec.net lan_ipv4=192.168.112.11
ci-dind ansible_host="{{ lan_ipv4 }}" lan_ipv4=192.168.111.10
```

**Comment by DaDummy** [comment]
Make sure `green.dashtec.net` is not reachable via the address/hostname in the variable `lan_ipv4` to have the playbook fail. Otherwise the issue might not become obvious.

**Comment by bcoca** [comment]
dupe of  #76007
fixed by #76017

**Comment by bcoca** [comment]
inventory:
```
all:
  hosts:
    delhost:
      hostvar1: 127.0.0.2
    orighost:
      hostvar1: 127.0.0.3
  vars:
    ansible_host: '{{hostvar1}}'
```
play
```
- hosts: all
  gather_facts: no
  tasks:
    - name: Retrieve file
      delegate_to: "delhost"
      shell: echo $SSH_CONNECTION
      ignore_errors: true
      register: connected

    - debug:
        msg: "{{ ' - '.join([ansible_host, hostvar1, connected.stdout]) }}"
```

correct otuput (using #76017 , both went to 127.0.0.2 the delegated to host: 
```
ok: [delhost] => {
    "msg": "127.0.0.2 - 127.0.0.2 - 127.0.0.1 60588 127.0.0.2 22"
}
ok: [orighost] => {
    "msg": "127.0.0.3 - 127.0.0.3 - 127.0.0.1 60588 127.0.0.2 22"
}

```

incorrect output (using devel/any stable, hostvar1 was templated using inventory_hostname instead of actual host used):
```
ok: [delhost] => {
    "msg": "127.0.0.2 - 127.0.0.2 - 127.0.0.1 60592 127.0.0.2 22"
}
ok: [orighost] => {
    "msg": "127.0.0.3 - 127.0.0.3 - 127.0.0.1 60656 127.0.0.3 22"
}

```
